Add pe_source module from pe-reports

aloftus23 opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

Move the existing pe-source module in [pe-reports](https://github.com/cisagov/pe-reports) to this repo.

## 💭 Motivation and context ##

Splitting the repo will allow other applications to install just the pe-source module without any of the report/email functionality in pe-reports.

## 🧪 Testing ##

Updates pytests and passes pre-commits.

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

aloftus23 committed:
Copy pe_source module for pe_reports repo

aloftus23 committed:
Create new setup.py and update import directories

aloftus23 committed:
Run linter

aloftus23 committed:
Fix pre-commit checks

aloftus23 committed:
Update README.md with examples

aloftus23 committed:
fix readme linter

aloftus23 committed:
Fix pytests

aloftus23 committed:
fix lint

aloftus23 committed:
Fix dnstwist script

aloftus23 committed:
fix lint

aloftus23 committed:
Don't tesst on python3.7 packages since it will fail with new package versions

For example, pandas 1.5.1

aloftus23 committed:
Also remove python 3.6

aloftus23 committed:
fix build.yml file

aloftus23 commented:
Ready for review

aloftus23 commented:
@dav3r This PR is ready. Let me know if you have any questions

dav3r commented on setup.py:
```diff
@@ -74,24 +74,40 @@ def get_version(version_file):
         "Programming Language :: Python :: 3",
-        "Programming Language :: Python :: 3 :: Only",
         "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
```
Since you removed support for Python 3.6 and 3.7, you should remove these lines:
```suggestion
```

aloftus23 replied:
Removed here: [0a5e6a4](https://github.com/cisagov/pe-source/pull/1/commits/0a5e6a4e477de2ae5e65480c87de40bde641ab5e)

dav3r commented on setup.py:
```diff
@@ -74,24 +74,40 @@ def get_version(version_file):
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
-        "Programming Language :: Python :: 3.10",
-        "Programming Language :: Python :: 3.11",
```
Why did you remove support for 3.10 and 3.11 here?  It looks like the [tests are passing](https://github.com/cisagov/pe-source/actions/runs/5592380198) for both of those versions.

aloftus23 replied:
That was a mistake grabbing code from the pe-reports repo. Added them back in here: [0a5e6a4](https://github.com/cisagov/pe-source/pull/1/commits/0a5e6a4e477de2ae5e65480c87de40bde641ab5e)

dav3r commented on setup.py:
```diff
@@ -74,24 +74,40 @@ def get_version(version_file):
+            "data/shodan/*",
+            "data/sixgill/*",
+            "data/dnsmonitor/*",
+            "data/pe_db/*",
```
These should be sorted alphabetically:
```suggestion
            "data/*",
            "data/dnsmonitor/*",
            "data/pe_db/*",
            "data/shodan/*",
            "data/sixgill/*",
```

dav3r commented on setup.py:
```diff
@@ -107,6 +123,10 @@ def get_version(version_file):
     },
-    # Conveniently allows one to run the CLI tool as `example`
-    entry_points={"console_scripts": ["example = example.example:main"]},
+    # Conveniently allows one to run the CLI tool as `pe-reports` or 'pe-mailer'
```
Correct a comment:
```suggestion
    # Conveniently allows one to run the CLI tool as `pe-source'
```

dav3r reviewed (changes requested):
I'll try to review the rest of this later this week.  I noted a few items that caught my eye.

aloftus23 committed:
Remove python 3.6/3.7 and add 3.10/3.11 to setup.py

aloftus23 committed:
Correct comment and alphabetize package_data

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

aloftus23 commented:
> I'll try to review the rest of this later this week. I noted a few items that caught my eye.

Thanks. Addressed your initial review. Will look out for the rest.

dav3r commented on .DS_Store:
This file should not be checked in to the repo.

aloftus23 replied:
Removed in [e25ecd0](https://github.com/cisagov/pe-source/pull/1/commits/e25ecd0655cc9dbd31841384c58a3bff40e8e187)

dav3r commented on .flake8:
```diff
@@ -22,4 +22,9 @@ select = C,D,E,F,W,B,B950
+# No longer complies with PEP8 See example, here:
+# https://github.com/PyCQA/pycodestyle/issues/373
+# and here: https://github.com/psf/black/issues/315
+ignore = E203,E501,E712,W503
```
The comment you wrote explains why you added E203 to the list of ignored rules, but it does not explain why you added [E712](https://www.flake8rules.com/rules/E712.html).

aloftus23 replied:
Addressed here [ccbeaac](https://github.com/cisagov/pe-source/pull/1/commits/ccbeaacc8ee8b5e59a624b48b86fa721399744d4). E712 isn't necessary for this repo so removed.

dav3r commented on .gitignore:
```diff
@@ -2,11 +2,20 @@
 .mypy_cache
 .pytest_cache
 .python-version
+.vscode
```
`.vscode` is not a Python-related file.  It should either be in a section called "IDE" or "VS Code".

aloftus23 replied:
Addressed in [f58ca92](https://github.com/cisagov/pe-source/pull/1/commits/f58ca923388e9cc50052a26ab644dc043e8e595c)

dav3r commented on README.md:
```diff
@@ -5,20 +5,66 @@
-for step-by-step instructions on how to start a new repository from
-a skeleton. This will save you time and effort when configuring a
-new repository!
+This package is used to gather and store data for the CISA Posture & Exposure Reports
+[P&E Reports](https://github.com/cisagov/pe-reports).
```
I'd format this a bit more nicely:
```suggestion
This package is used to gather and store data for the CISA
[Posture & Exposure Reports](https://github.com/cisagov/pe-reports).
```

dav3r commented on README.md:
```diff
@@ -5,20 +5,66 @@
+                                    orgs in the pe database. Orgs in the list must match the
+                                    IDs in the cyhy-db. E.g. DHS,DHS_ICE,DOC
+                                    [default: all]
+  -csg --cybersix-methods=METHODS   A comma-separated list of cybersixgill methods to run.
```
Typically, short flag options are a only single character.  Also, use consistent capitalization:
```suggestion
  -c --cybersix-methods=METHODS     A comma-separated list of Cybersixgill methods to run.
```

dav3r commented on setup.py:
```diff
@@ -42,19 +42,19 @@ def get_version(version_file):
     project_urls={
-        "Source": "https://github.com/cisagov/pe-source",
-        "Tracker": "https://github.com/cisagov/pe-source/issues",
+        "Source": "https://github.com/cisagov/pe-reports",
+        "Tracker": "https://github.com/cisagov/pe-reports/issues",
```
I'm not sure why these were changed:
```suggestion
        "Source": "https://github.com/cisagov/pe-source",
        "Tracker": "https://github.com/cisagov/pe-source/issues",
```

jsf9k replied:
I agree.  This is very confusing, and these files are definitely in the cisagov/pe-source repo.

dav3r commented on src/pe_source/__init__.py:
```diff
@@ -0,0 +1,28 @@
+DEBUG = False
+
+# Setup Logging
+"""Set up logging and call the run_pe_script function."""
```
I don't see a call to `run_pe_script` - am I missing something or should this doc string be modified?
```suggestion
"""Set up logging."""
```

dav3r commented on src/pe_source/data/database.ini:
The sections in this file should be alphabetized.

aloftus23 replied:
Fixed here: [d67d2d7](https://github.com/cisagov/pe-source/pull/1/commits/d67d2d7b75be430694371f138036c8a1659e2e97)

dav3r commented on tests/.DS_Store:
This file should not be checked in.

aloftus23 replied:
Removed here: [e63fb27](https://github.com/cisagov/pe-source/pull/1/commits/e63fb27037db798bf0bee8092bfcd187ed4bb01f)

dav3r reviewed (changes requested):
I made it a little bit further through my review, but there's a lot of code in this PR.  I'll have to continue at a later date.

dav3r commented on src/pe_source/pe_scripts.py:
```diff
@@ -0,0 +1,129 @@
+"""A tool for gathering pe source data.
+
+Usage:
+    pe-source DATA_SOURCE [--log-level=LEVEL] [--orgs=ORG_LIST] [--cybersix-methods=METHODS] [--soc_med_included]
```
Is there a reason this file is named `pe_scripts.py` and not `pe_source.py`?  It just seems unnecessarily confusing to this outside observer, but if there's a good reason, I'm interested in hearing it.

aloftus23 replied:
Good point. Changed it to pe_source here: [b23a404](https://github.com/cisagov/pe-source/pull/1/commits/b23a40403758aa79a078b491e0072cf2f175edcf)

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -111,14 +111,12 @@ jobs:
           - "3.10"
           - "3.11"
         include:
           - os: ubuntu-20.04
```
I think these two lines can be deleted:
```suggestion
```

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -207,14 +205,12 @@ jobs:
           - "3.10"
           - "3.11"
         include:
           - os: ubuntu-20.04
```
```suggestion
```

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -260,14 +256,12 @@ jobs:
           - "3.10"
           - "3.11"
         include:
           - os: ubuntu-20.04
```
```suggestion
```

jsf9k commented on .gitignore:
```diff
@@ -2,11 +2,20 @@
+pe_reports_logging.log
+src/pe_source/data/dnstwist_output.txt
+
+
```
Remove unnecessary blank line:
```suggestion
```

aloftus23 replied:
Addressed here [f58ca92](https://github.com/cisagov/pe-source/pull/1/commits/f58ca923388e9cc50052a26ab644dc043e8e595c)

jsf9k commented on .pre-commit-config.yaml:
```diff
@@ -116,15 +116,28 @@ repos:
+          - types-PyYAML
+          - types-requests
           - types-setuptools
+          - types-python-dateutil==2.8.19
```
Why not put this dependency in alphabetical order like the rest?

aloftus23 replied:
Fixed here: [5d1b723](https://github.com/cisagov/pe-source/pull/1/commits/5d1b723a956b0f71fc8982c98f82b1553c3a8517)

jsf9k commented on README.md:
```diff
@@ -5,20 +5,66 @@
+[P&E Reports](https://github.com/cisagov/pe-reports).
+
+Data of interest include *Exposed Credentials, Domain Masquerading, Malware,
+Inferred Vulnerabilities and the Dark Web*. The data collected for the reports
```
Oxford commas make lists less prone to misunderstanding:
```suggestion
Inferred Vulnerabilities, and the Dark Web*. The data collected for the reports
```

jsf9k commented on README.md:
```diff
@@ -5,20 +5,66 @@
+
+- Add database/API credentials to `src/pe_source/data/pe_db/database.ini`
+
+- `pip install -e .`
```
Remove unnecessary blank lines:
```suggestion
- `git clone https://github.com/cisagov/pe-source.git`
- Add database/API credentials to `src/pe_source/data/pe_db/database.ini`
- `pip install -e .`
```

jsf9k commented on README.md:
````diff
@@ -5,20 +5,66 @@
+
+```console
+Usage:
+    pe-source DATA_SOURCE [--log-level=LEVEL] [--orgs=ORG_LIST] [--cybersix-methods=METHODS] [--soc_med_included]
````
The `--soc_med_included` flag is not listed under "Options:" below.

Also, the other options use `-` instead of `_`.  I suggest renaming the option `--soc-med-included` for consistency.

aloftus23 replied:
Fixed here: [d8e46a6](https://github.com/cisagov/pe-source/pull/1/commits/d8e46a6c95a0d185990945b7ba9d347bdeb1524c)

jsf9k commented on setup.py:
```diff
@@ -74,24 +74,40 @@ def get_version(version_file):
         "Programming Language :: Python :: 3.9",
         "Programming Language :: Python :: 3.10",
         "Programming Language :: Python :: 3.11",
-        "Programming Language :: Python :: Implementation :: CPython",
```
I don't think this line should have been removed.  The `psycopg2-binary` package is almost certainly compiled against CPython.

aloftus23 replied:
Added back in [ac7c6ea](https://github.com/cisagov/pe-source/pull/1/commits/ac7c6ea02482c2a46d5685fb1d618f619b5c0b7a)

jsf9k commented on src/pe_source/__init__.py:
```diff
@@ -0,0 +1,28 @@
+# Standard Python Libraries
+import logging
+
+from ._version import __version__  # noqa: F401
```
Move the comment to just before the line it refers to:
```suggestion
# Standard Python Libraries
import logging

# We disable a Flake8 check for "Module imported but unused (F401)" here because
# although this import is not directly used, it populates the value
# package_name.__version__, which is used to get version information about this
# Python package.
from ._version import __version__  # noqa: F401
```

jsf9k commented on setup.py:
```diff
@@ -74,24 +74,40 @@ def get_version(version_file):
     python_requires=">=3.6",
     # What does your project relate to?
-    keywords="skeleton",
+    keywords="posture and exposure source",
```
`And` is probably not intended as a keyword:
```suggestion
    keywords="posture exposure source",
```

jsf9k commented on src/pe_source/pe_scripts.py:
```diff
@@ -0,0 +1,129 @@
+"""A tool for gathering pe source data.
+
+Usage:
+    pe-source DATA_SOURCE [--log-level=LEVEL] [--orgs=ORG_LIST] [--cybersix-methods=METHODS] [--soc_med_included]
```
Underscores are not used for any of the other options:
```suggestion
    pe-source DATA_SOURCE [--log-level=LEVEL] [--orgs=ORG_LIST] [--cybersix-methods=METHODS] [--soc-med-included]
```

aloftus23 replied:
Fixed here: [d8e46a6](https://github.com/cisagov/pe-source/pull/1/commits/d8e46a6c95a0d185990945b7ba9d347bdeb1524c) and [3a2e262](https://github.com/cisagov/pe-source/pull/1/commits/3a2e2620cbe58191f1652fca4b3adcfe78db6ba8)

jsf9k commented on src/pe_source/pe_scripts.py:
```diff
@@ -0,0 +1,129 @@
+                                    If not specified, all will run. Valid values are "alerts",
+                                    "credentials", "mentions", "topCVEs". E.g. alerts,mentions.
+                                    [default: all]
+  -sc --soc_med_included            Include social media posts from cybersixgill in data collection.
```
Underscores are not used for any of the other options:
```suggestion
  -sc --soc-med-included            Include social media posts from cybersixgill in data collection.
```

aloftus23 replied:
Fixed in [d8e46a6](https://github.com/cisagov/pe-source/pull/1/commits/d8e46a6c95a0d185990945b7ba9d347bdeb1524c)

dav3r commented on src/pe_source/cybersixgill.py:
```diff
@@ -0,0 +1,407 @@
+    top_cves,
+)
+
+# Set todays date formatted YYYY-MM-DD and the start_date 30 days prior
```
This section doesn't include any date formatting:
```suggestion
# Set todays date and the start_date 30 days prior
```

dav3r commented on src/pe_source/cybersixgill.py:
```diff
@@ -0,0 +1,407 @@
+        list = ""
+        for pe_org in pe_orgs:
+            list = list + pe_org["cyhy_db_name"] + ","
+        print(list)
```
This output should be logged as either debug or info, rather than printed.

Also, you can get rid of that temporary `list` variable with something like this:
```suggestion
        LOGGER.info(",".join([org["cyhy_db_name"] for org in pe_orgs]))
```

jsf9k commented on src/pe_source/data/dnsmonitor/source.py:
```diff
@@ -0,0 +1,78 @@
+    for org_index, org_row in org_names_df.iterrows():
+        for domain_index, domain_row in domain_df.iterrows():
+            if org_row["domain_name"] == domain_row["domainName"]:
+                domain_df.at[domain_index, "org"] = org_row["org"]
```
You could possibly rewrite this to avoid the double `for` loop and speed it up.  Just something to think about.

aloftus23 replied:
Good call. Made adjustments here: [c362c8e](https://github.com/cisagov/pe-source/pull/1/commits/c362c8e64be4a7e20b478e3e37828fdc2316122c)

jsf9k commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,562 @@
+import scrubadub
+import scrubadub.detectors.date_of_birth
+
+# List of unique regexes to identify each state's Drivers License format in a larger string
```
Are all 50 states not listed because the states not listed use a person's SSN as the identifier?

jsf9k commented on src/pe_source/data/shodan/shodan_search.py:
```diff
@@ -0,0 +1,420 @@
+    for i, ip_chunk in enumerate(ip_chunks):
+        count = i + 1
+        try_count = 1
+        while try_count < 7:
```
Magic numbers like `7` should be avoided; use variables so you can easily change the value later if necessary.

aloftus23 replied:
Fixed here: [10f0e6d](https://github.com/cisagov/pe-source/pull/1/commits/10f0e6d6e5c97425615bd3c057aa9040ecbf8928)

jsf9k reviewed (changes requested):
I have some change requests and suggestions for improvement.

dav3r commented on src/pe_source/cybersixgill.py:
```diff
@@ -0,0 +1,407 @@
+                count += 1
+                # Get sixgill_org_id associated with the PE org
+                try:
+                    sixgill_org_id = sixgill_orgs[org_id][5]
```
This "5" magic number should have a comment explaining it.

aloftus23 replied:
Fixed here: [10f0e6d](https://github.com/cisagov/pe-source/pull/1/commits/10f0e6d6e5c97425615bd3c057aa9040ecbf8928)

dav3r commented on src/pe_source/cybersixgill.py:
```diff
@@ -0,0 +1,407 @@
+        """Get alerts."""
+        LOGGER.info("Fetching alert data for %s.", org_id)
+        soc_med_platforms = [
+            "twitter",
+            "Twitter",
+            "reddit",
+            "Reddit",
+            "Parler",
+            "parler",
+            "linkedin",
+            "Linkedin",
+            "discord",
+            "forum_discord",
+            "raddle",
+            "telegram",
+            "jabber",
+            "ICQ",
+            "icq",
+            "mastodon",
```
For ease of maintenance, can this list be sorted alphabetically?
```suggestion
            "discord",
            "forum_discord",
            "icq",
            "ICQ",
            "jabber",
            "linkedin",
            "Linkedin",
            "mastodon",
            "parler",
            "Parler",
            "raddle",
            "reddit",
            "Reddit",
            "telegram",
            "twitter",
            "Twitter",
```

dav3r commented on src/pe_source/data/dnsmonitor/root_domains_dnsmonitor.csv:
This large CSV should be sorted by domain in order to make future updates easier.

aloftus23 replied:
Good idea. Fixed here: [53e8e28](https://github.com/cisagov/pe-source/pull/1/commits/53e8e280d2c5a27045ef8b4ca6da9dd6823a4439)

dav3r commented on src/pe_source/data/dnsmonitor/source.py:
```diff
@@ -0,0 +1,78 @@
+
+    # A
+    try:
+        ip_address = str(socket.gethostbyname(dom_perm))
+        if ":" in ip_address:
+            ipv6 = ip_address
+            ipv4 = ""
+        else:
+            ipv4 = ip_address
+            ipv6 = ""
+    except Exception:
+        ipv4 = ""
+        ipv6 = ""
```
The Python built-in [`ipaddress` library](https://docs.python.org/3/library/ipaddress.html) should be used here to determine v4 vs. v6 IP addresses.

aloftus23 replied:
Fixed here: [37e73ba](https://github.com/cisagov/pe-source/pull/1/commits/37e73bae09e4c08dffd9ad953ec3f1a0a6933421)

dav3r reviewed (commented):
Got through a few more files - will continue to chip away at this review when I am able to.

aloftus23 committed:
Remove E712 ignore

Was pulled from another repo. Not needed here

aloftus23 committed:
Change pe_scripts to pe_source

aloftus23 committed:
Delete .DS_Store

Remove .DS_Store

aloftus23 committed:
Fix .gitignore

aloftus23 committed:
Fix formatting in README

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>
Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

aloftus23 committed:
Fix setup.py mistakes

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>
Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

aloftus23 committed:
Fix commenting in __init__

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>
Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

aloftus23 committed:
Remove ubuntu-10.04 from build.yml

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

aloftus23 committed:
Alphabetize .pre-commit configs

aloftus23 committed:
Merge branch 'AL-copy-pe-reports' of https://github.com/cisagov/pe-source into AL-copy-pe-reports

aloftus23 committed:
Fix options in docopt and readme

Change --soc_med_included to --soc-med-included

change -csg to -c and -sc to -s to consistently keep the flags one character

aloftus23 committed:
Alphabetize database.ini

aloftus23 committed:
Delete tests/.DS_Store

aloftus23 committed:
Re-add CPython to setup.oy

aloftus23 committed:
Fix comment in src/pe_source/cybersixgill.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

aloftus23 committed:
Simplify pe_org log in src/pe_source/cybersixgill.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

aloftus23 committed:
Run linter

aloftus23 committed:
Merge branch 'AL-copy-pe-reports' of https://github.com/cisagov/pe-source into AL-copy-pe-reports

aloftus23 committed:
Use ipaddress python library to determine v6 vs v4

aloftus23 committed:
Sort root_domains_dnsmonitor by domain

aloftus23 committed:
Sort cybersixgill.py sites

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

aloftus23 committed:
Fix docopt for soc-med-included

aloftus23 committed:
Merge branch 'AL-copy-pe-reports' of https://github.com/cisagov/pe-source into AL-copy-pe-reports

aloftus23 committed:
Address magic numbers

Add comment explaining for cybersixgill.py and add variable for retry count

aloftus23 committed:
Replace double for-loop with merge in dnsmonitor/source
